File: eternal_return/src/main/java/org/EternalReturn/ERPlayer/ERPlayer.java
package org.EternalReturn.ERPlayer;

import org.EternalReturn.ERCharacter.character.NullCharacter;
import org.EternalReturn.ERCharacter.ERCharacter;
import org.EternalReturn.ERPlayer.Gui.Inventory.UpgradeSystem.UpgradeGuiController;
import org.EternalReturn.ERPlayer.Gui.Inventory.UpgradeSystem.View.UpgradeGui;
import org.EternalReturn.util.Gui.InventoryGui.View.IController;
import org.EternalReturn.util.Geometry.MathVector.Vec3d;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

public class ERPlayer {

    private Player player;

    private UpgradeGui upgradeGui;
    private IController upgradeGuiController;
    private ERCharacter character;

    public void free(){
        upgradeGuiController.free();
        upgradeGui.free();
    }

    public ERPlayer(Player p){
        player = p;
        upgradeGui = new UpgradeGui(p);
        upgradeGuiController = new UpgradeGuiController(this, upgradeGui);
        character = new NullCharacter(this);
    }

    //getter
    public Player getPlayer() {
        return player;
    }

    public IController getUpgradeGuiController(){
        return upgradeGuiController;
    }


    @NotNull public ERCharacter getCharacter(){
        return this.character;
    }

    //setter
    public UpgradeGui getUpgradeGui() {
        return upgradeGui;
    }

    public void sendMessage(String str) {
        player.sendMessage(str);
    }

    public Vec3d getPos(){
        return new Vec3d(player.getLocation());
    }

    public void setCharacter(@NotNull ERCharacter character){
        this.character = character;
    }

}

File: eternal_return/src/main/java/org/eternalreturn/erplayer/ERPlayer.kt
package org.eternalreturn.erplayer

import org.bukkit.entity.Entity
import org.bukkit.entity.Player
import org.eternalreturn.ercharacter.ERCharacter
import org.eternalreturn.system.EREngine
import org.eternalreturn.util.Gui.InventoryGui.View.IController

open class ERPlayer(p: Entity, engine: EREngine) : ERCharacter(p, engine) {

    var attackCooldown : Float = 0.0F;
    val player : Player
        get() {
            return entity as Player
        }

    fun sendMessage(str: String) {
        entity.sendMessage(str)
    }

    override fun shootRay(){
        super.shootRay();
        attackCooldown = player.attackCooldown;
    }

    //fun playSound()

    override val name: String
        get() = "NullPlayer"
}
